src/wasm/wasm-s-parser.cpp

```diff
@@ -2680,12 +2689,21 @@ Expression* SExpressionWasmBuilder::makeArrayNew(Element& s, bool default_) {
   return Builder(wasm).makeArrayNew(rtt, size, init);
 }
 
-Expression* SExpressionWasmBuilder::makeArrayGet(Element& s, bool signed_) {
+Expression* SExpressionWasmBuilder::makeArrayGet(Element& s, Packed packed) {
   auto heapType = parseHeapType(*s[1]);
+  auto element = heapType.getArray().element;
+  if (element.isPacked() && packed == Packed::Non) {
+    throw ParseException(
+      "packed element must be read as packed", s.line, s.col);
+  }
+  if (!element.isPacked() && packed != Packed::Non) {
+    throw ParseException(
+      "non-packed element must not be read as packed", s.line, s.col);
+  }
```

Comment:
This check could be factored out into a shared function.

Reply:
Fair enough. The error text is different, though, which will make that slightly less nice, but I agree it's probably better overall.